server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,463 @@
+    ) {
+        String msg = null;
+        if (awarenessAttributes == null) {
+            msg = "awareness attribute not set to the cluster.";
+        } else if (forcedAwarenessAttributes == null) {
+            msg = "forced awareness attribute not set to the cluster.";
+        } else if (awarenessAttributes.contains(decommissionAttribute.attributeName()) == false) {
+            msg = "invalid awareness attribute requested for decommissioning";
+        } else if (forcedAwarenessAttributes.containsKey(decommissionAttribute.attributeName()) == false) {
+            msg = "forced awareness attribute [" + forcedAwarenessAttributes.toString() + "] doesn't have the decommissioning attribute";
+        } else if (forcedAwarenessAttributes.get(decommissionAttribute.attributeName())
+            .contains(decommissionAttribute.attributeValue()) == false) {
+                msg = "invalid awareness attribute value requested for decommissioning. Set forced awareness values before to decommission";
+            }
+
+        if (msg != null) {
+            throw new DecommissioningFailedException(decommissionAttribute, msg);
+        }
+    }
+
+    private static void ensureEligibleRequest(
+        DecommissionAttributeMetadata decommissionAttributeMetadata,
+        DecommissionAttribute requestedDecommissionAttribute
+    ) {
+        String msg = null;
+        if (decommissionAttributeMetadata != null) {
+            // check if the same attribute is registered and handle it accordingly
+            if (decommissionAttributeMetadata.decommissionAttribute().equals(requestedDecommissionAttribute)) {
+                switch (decommissionAttributeMetadata.status()) {
+                    // for INIT and FAILED - we are good to process it again
+                    case INIT:
+                    case FAILED:
+                        break;
+                    case IN_PROGRESS:
+                    case SUCCESSFUL:
+                        msg = "same request is already in status [" + decommissionAttributeMetadata.status() + "]";
+                        break;
+                    default:
+                        throw new IllegalStateException(
+                            "unknown status [" + decommissionAttributeMetadata.status() + "] currently registered in metadata"
+                        );
+                }
+            } else {
+                switch (decommissionAttributeMetadata.status()) {
+                    case SUCCESSFUL:
+                        // one awareness attribute is already decommissioned. We will reject the new request
+                        msg = "one awareness attribute ["
+                            + decommissionAttributeMetadata.decommissionAttribute().toString()
+                            + "] already successfully decommissioned, recommission before triggering another decommission";
+                        break;
+                    case IN_PROGRESS:
+                    case INIT:
```

Comment:
@shwetathareja my suggestion was to annotate the request with a `continuation` flag once we are processing the cluster state update and know that the current request is the one that has been able to acquire the lock. So before marking the state as `INIT` we could potentially set that attribute to prevent other threads from executing concurrently. Anytime the leader switch happens the same request would automatically get retried and the same `continuation` flag be effectively making the lock `re-entrant`. Only when the thread processing the request dies will the request be stuck in an INIT preventing other requests from rescing it. But since this should be rare and unexpected for the thread to unexpectedly die or atleast rare for, we have the `DELETE` API to our rescue which should clean the decommissioned state and allow a new request to start afresh.

Reply:
>As, decommissioning flow starts when there is an external API trigger and lets say if master switches or it is restarted then the decommissioning which was in progress could be abandoned. Now the decommissioning request could stay in that state forever until the request is retried again by the user. This behavior is as per design. But we need to ensure is that when request is retried it starts from the start and continue from where it left off. The previous executed steps shouldn't fail (just be no-op).

1. Master switches is already handled for INIT step and the already existing retry mechanism of TransportClusterManagerNodeAction ensures that until we explicitly send the response to the listener. 
2. Now for IN_PROGRESS request it become tricky. IN_PROGRESS comes in picture when abdication was successful and to-be-decommissioned leader eligible nodes were kicked out from VotingConfig. Also note here that we have responded back to the original listener in this state. Now, if it is an abandoned IN_PROGRESS request or a fresh one, we cannot say with guarantee. Hence, just forwarding the new request and not blocking it in beginning could just be a add on to the cluster which is already under stress and also could interfere with existing `IN_PROGRESS` if it is not abandoned. User can always proceed to wipe off the metadata to restart a new request

Reply:
> With both INIT/ IN PROGRESS - user should be allowed to retry. There is a catch that there could be multiple concurrent executions as well. To ensure correctness each step is already idempotent. To ensure only one active execution should proceed and other should early return, adding an extra attribute like monotonically increasing field like version can help (which is incremented for a new execution trigger via API call). And any changes to decommissionMetadata also uses version for optimistic concurrency control. Also this will help during race condition arising from multiple concurrent executions. e.g. Execution1 intend to change the change from IN_PROGRESS to FAILED and EXECUTION2 from IN_PROGRESS to SUCCESS but it will depend on which enqueued first instead of which was latest execution.

This does make sense to me. Versioning of the metadata could certainly help with concurrent executions. For scope of this PR do you think we can take this as a follow up item and implemented in a follow up PR?

Reply:
> @shwetathareja my suggestion was to annotate the request with a continuation flag once we are processing the cluster state update and know that the current request is the one that has been able to acquire the lock. So before marking the state as INIT we could potentially set that attribute to prevent other threads from executing concurrently. Anytime the leader switch happens the same request would automatically get retried and the same continuation flag be effectively making the lock re-entrant

@Bukhtawar retry mechanism is limited to INIT step as once INIT is completed, we send the response back to the user and hence system will not be able to retry itself until the user initiates it

Reply:
Adding the INIT behaviour below in the diagram. INIT in itself is made idempotent and ensures no two thread gives out different results. Last thread executing the INIT will be taken as source of truth. Also, to add here - it is guaranteed that when this entire execution starts, no new node with decommissioned attribute will be able to join the cluster, which ensures all threads running for INIT will attempt to achieve the same result and prevents us from cases where one thread executing INIT is now seeing a to-be-decommissioned leader and other one not 

![Leader Decommission jpeg](https://user-images.githubusercontent.com/25440813/191617485-0ec098ad-0e64-4b30-81e2-473345b5c2b6.jpeg)

Reply:
In order to the simplify the further (and taking into consideration the above conversations). We can look into following option during the follow up PRs (those would be blocker to decommission feature launch)
1. A new proposal - Both setting the decommission metadata with INIT state and voting exclusion can be done in a single cluster state update as atomic operation instead of separate transport call for exclusion. The decommission request would only do this change in sync and rest of the operations including clearing of voting config can happen in the background. This would simplify the code significantly. No need to worry about concurrent executions.
2. Like @Bukhtawar suggested to have a flag in the request object which indicates whether it is retry or new request so that INIT->INIT can be allowed. All concurrent requests are rejected. In case, decommissioning gets stuck at INIT or IN_PROGRESS, user would first clear the metadata using delete API and then trigger a new decommissioning request.
3. Support metadata versioning in case we wish to support concurrent requests. I dont think we need to support concurrent requests.

Reply:
Please open follow up github issues and add their reference.

Reply:
Opened 2 issues to track above suggested changes - 
#4543 - to gracefully handle concurrent request. Here we can discuss on versioning, allowing concurrent request, retry request with a flag, etc
#4571 - To merge INIT and voting config exclusion in a single state update call

Reply:
Tracking all the meta issues here - https://github.com/opensearch-project/OpenSearch/issues/3957